### DISCUSSION TITLE
Could target-typed new be updated to never match a pointer type?

### DISCUSSION THREAD

**Discussion by jnm2** [dotnet/csharplang#4895] [Language Ideas]
This would only affect the behavior of code that did not previously compile.

I think overload resolution should first consider the set of overloads that don't have a pointer type for the `new()` argument, since resolving to any of those is guaranteed to cause CS1919. Overload resolution would then succeed in the example below.

```cs
struct SomeStruct
{
    void M()
    {
        // These errors currently disappear if the SomeStruct* overload is commented out
        
        // ❌ CS1919 Unsafe type 'SomeStruct*' cannot be used in object creation
        // ❌ CS0214 Pointers and fixed size buffers may only be used in an unsafe context
        M2(new() { /* ... */ });
    }

    // Overloads like this are generated by Microsoft.Windows.CsWin32
    void M2(in SomeStruct _) { }
    unsafe void M2(SomeStruct* _) { }
}
```

Overloads like this are an exceedingly common pattern generated by CsWin32. CsWin32 is a new official vehicle for the Windows SDK for C# developers (https://blogs.windows.com/windowsdeveloper/2021/01/21/making-win32-apis-more-accessible-to-more-languages/).

**Comment by RikkiGibson** [comment]
Overload resolution is somewhat difficult to change. It feels like the request is for the argument `new()` to somehow contribute the fact that "it is not allowed to be of an unsafe type" to overload resolution. The main questions I have are:
1. How can this be specified?
2. What is the implementation and test burden?

**Reply by jnm2** [reply]
Neither of those are questions I know how to answer right away, but I might be able to take a stab at the first one. Do you think that attempt could prove useful to you?

**Reply by 333fred** [reply]
Personally, no. I wouldn't be interested in this feature. I'd be much more interested in the ranked overload feature to resolve otherwise ambiguous overloads.

**Reply by Rekkonnect** [reply]
Are we aware of any other scenarios that could result in an ambiguous overload resolution including a managed and an unsafe type?

**Reply by RikkiGibson** [reply]
[SharpLab](https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetE+UkALIgLIAMAFAJTa5Yy4MnkB2ArgDbuUDciA9H0QBTAE4iA9iLQBlUhEgAuRABUAFkMQBjAIadEASwDOibQFsARvoDmrca2PmhAFwDuQoc0RP1iAGbjOcRd9ZitEU2dVcQATY0lEAAcJBNEnfSFDJQByVAA6YnIQpwAqSiyTZmjEHPzycXMAKyFNJzKaHABfdsRuhERWZkNtXw1CJiLixKcRSkxurvpcPrGC+qaWxDXZugYFjqA===). Here a method with pointer type parameters causes an ambiguity error even though it isn't allowed to be called. The goal might be to prevent the results of overload resolution changing depending on the unsafe context.

```cs
public class C {
    public void M0()
    {
        M(null); // error CS0121: The call is ambiguous between the following methods or properties: 'C.M(int*)' and 'C.M(object)'
    }
    
    public unsafe void M(int* ptr) {
    }
    public void M(object obj) {
    }
}
```
